backend/compact-connect/lambdas/staff-users/tests/unit/test_collect_changes.py

```diff
@@ -0,0 +1,161 @@
+from tests import TstLambdas
+
+
+class TestCollectChanges(TstLambdas):
+    """
+    Testing that permissions changes are parsed correctly from the API
+    """
+
+    def test_compact_changes(self):
+        from utils import collect_changes
+
+        resp = collect_changes(
+            path_compact='aslp',
+            scopes={'openid', 'email', 'aslp/admin', 'aslp/aslp.admin'},
+            compact_changes={
+                'actions': {
+                    'admin': True,
+                    'read': False
+                },
+                'jurisdictions': {}
+            }
+        )
+        self.assertEqual(
+            {
+                'compact_action_additions': {'admin'},
+                'compact_action_removals': {'read'},
+                'jurisdiction_action_additions': {},
+                'jurisdiction_action_removals': {}
+            },
+            resp
+        )
+
+    def test_jurisdiction_changes(self):
+        from utils import collect_changes
+
+        resp = collect_changes(
+            path_compact='aslp',
+            scopes={'openid', 'email', 'aslp/admin', 'aslp/oh.admin'},
+            compact_changes={
+                'jurisdictions': {
+                    'oh': {
+                        'actions': {
+                            'admin': True,
+                            'write': False
+                        }
+                    }
+                }
+            }
+        )
+        self.assertEqual(
+            {
+                'compact_action_additions': set(),
+                'compact_action_removals': set(),
+                'jurisdiction_action_additions': {
+                    'oh': {'admin'}
+                },
+                'jurisdiction_action_removals': {
+                    'oh': {'write'}
+                }
+            },
+            resp
+        )
+
+    def test_disallowed_jurisdiction_changes(self):
+        from exceptions import CCAccessDeniedException
+        from utils import collect_changes
+
+        with self.assertRaises(CCAccessDeniedException):
+            collect_changes(
```

Comment:
Doesn't `collect_changes` just produce a new object, in that it's a pure function? I'd expect the function to throw these errors if it was actually _making_ the changes in the DB, but not just if it were creating an object for updating. (This might just be a design decision you made, which is totally fine. Just clarifying my understanding.)

Reply:
The design choice I made was to tightly couple access control with the logic that interprets the changes, so that it would be harder to accidentally leave an opening in the permission model compared to if the two evaluations were independent. In the case that one of the changes identified is not allowed based on the provided scopes, the function will raise an exception. I should probably rename this function to make that behavior more clear.

Reply:
Got it! Thanks for the explanation. Makes sense.